next `solid-three`

bigmistqke opened the pull request:
This branch continues the `context-and-proxies`-branch from @vorth in which he took the proxy-implementation from @nksaraf `solid-three-chess` branch of `react-three-fiber`.

This branch' intention is to remove all the dependencies and artefacts from `react`, remove the dependency on `zustand`  as an internal state manager in favour of `solid/store` and streamline the code, while keeping as close as possible to r3f's v9 codebase.

## what has been done
- fixed the broken event-system
- remove all references to useRef `.current`
- remove dependency on zustand
    - `useThree` works different then r3f's, while they have selectors to subscribe to store-updates, with our `useThree` is just a wrapper around `useContext(ThreeContext)`, as we don't have to worry about unnecessary re-renders.
    - a lot of code in `core` now uses solid-store specific code to update the store, so the distinction between `core` and `solid` directory is further made unclear.
    - it simplified some of the internal code, since we don't have to `const state = store.getState()` and the like.
    - `store.subscribe()` has been replaced by `createEffect(on(() => store, () => {…}))` which I am not 100% sure is the same. `store. subscribe` might be deep-tracking bc off the whole immutable-thing.
- I added an optimization to `applyProps` to prevent unnecessary creations/cleanups of effects with `mapArray(() => Object.keys(props), ...)` and checking if prop is possibly reactive (`=== a getter or function`)
- refactors/cleanups:
    -  the code in `solid/index.ts` is moved to `solid/renderer.ts`.  `core/renderer.ts` is removed. It contained some types it itself had inherit from the removed reconciler.ts, they have been moved to `three-types.ts`.
    - `index.ts`, ~`solid/index.ts`~ _(directories are merged, see below)_ and `core/index.ts` now only export files
    - removed secondary implementation of `applyProps`
    - removed unused three-types (remains of r3f's main branch)
    - replaced our custom three-types with r3f's v9 branch' implementation (it used to be done manually, but in the meanwhile they also figured out how to infer it from `three.js`)
    - our `.prettierrc` had the line `"plugins": ["prettier-plugin-tailwindcss"]` which caused my prettier to not function. I replaced our `.prettierrc` with the one of r3f to prevent unnecessary differences between codebases.
    - went ahead and merged the `core` and `solid`-repo back into `core`, more closely mimicking r3f's project-structure
    - renamed `core/components` to `core/proxy` (bc I thought it was a better parallel with `reconciler` and i preferred how it looked lol)
    - refactored the codebase to mimick r3f's codebase as close as possible
    
## reflections

~Currently our codebase is somewhere between r3f's main branch and v9. For example, `Instance` is typed like the main branch, but we have `Stages` like v9. Probably these changes happened after nikhil made his implementation [(found pr when they made the change)](https://github.com/pmndrs/react-three-fiber/pull/2465). Idk if we should pull everything closer to v9 or not. _I think yes, but my previous attempts have all failed._~ 

_Instance has been ported to current v9 implementation. Our codebase is now up-to-date._

~As I said above; since we now have solid-store-related code throughout the codebase idk if it still makes sense to make the distinction between the two. I think mb it's best to just bring it back into 1 repo and try to mimic the structure as closely to r3f as possible, so we can at least manually compare the two when needed.~

_I went ahead and merged those directories_

I think it would probably be good to try out some more complex examples, to test out this version. Maybe port the examples of r3f.

vorth committed:
Switched from universal renderer to context and T proxy

This commit discards Nikhil Saraf's older architecture based on the SolidJS
universal renderer, in favor of his newer approach that looks more like Threlte:

https://github.com/nksaraf/react-three-fiber/blob/chess-solid-three/solid-example/src/components/ModernPiece.tsx

This approach uses context to maintain the scene graph, thus avoiding the need
for the universal renderer, which adds build-time complexity and a layer (or two)
of abstraction.  It also uses a "T" proxy object to instantiate the three.js objects
and add them to the scene appropriately.

Issues:

The example seems to leave the canvas at some default size, rather than filling the
viewport.

This approach still has a dependency on "scheduler", an NPM package apparently
created as a reusable React internal package.  We might consider replacing that.

bigmistqke committed:
(fix) core/events add back prepareRay

bigmistqke committed:
(fix) comment out the resetting of raycasters

unclear why this is being done in the first place.

bigmistqke committed:
(fix) extend state.events with events(store) instead of overwriting

in core/events.ts the raycasting was checking for state.events.enabled, but this was being overwritten with events().

bigmistqke committed:
(fix) add global css to example for fullscreen view

bigmistqke committed:
(fix) add updates from r3f's v9 branch

bigmistqke committed:
remove unused functions in renderer

bigmistqke committed:
replace React.MutableRefObject with temp MutableRefObject type

bigmistqke committed:
refactor Subscription.ref as Callback instead of MutableRefObject<Callback>

remove artefact from original useRef

bigmistqke committed:
refactor store.internal.lastEvent as DomEvent | null instead of MutableRefObject<DomEvent | null>

remove artefact from original useRef

bigmistqke committed:
(fix) remove unused i declaration

bigmistqke committed:
cleanup solid/hooks

bigmistqke committed:
(fix) remove MutableRefObject in internal.subscribe()

bigmistqke committed:
(fix) remove .current in frameCallback

bigmistqke committed:
remove unused three-types

bigmistqke committed:
add utilit-types to peer-dep

bigmistqke committed:
simplify vite.config.ts

bigmistqke committed:
remove pnpm-workspace.yaml

bigmistqke committed:
update pnpm-lock.yaml

bigmistqke committed:
update example

bigmistqke committed:
remove scheduler from dependencies

bigmistqke committed:
add solid devtools and change to type: module

bigmistqke committed:
initial commit to port from zustand to solid-store

bigmistqke committed:
remove zustand.ts and zustand from deps

bigmistqke committed:
remove rootState.get since it was redundant

bigmistqke committed:
remove all `const state = store`s

bigmistqke committed:
(fix) update package.json

bigmistqke committed:
add get()-back to get back typings for rootState

circular dependency was causing type-errors

bigmistqke committed:
(fix) `RangeError: Maximum Call Stack Size Exceeded`-bug

An endless effect was created by calling `props.onCreated` inside a useIsomorphicLayoutEffect.
In the original r3f-code it is an IsomorphicLayoutEffect with empty dependency-array, but using onMount did not render any image.

bigmistqke committed:
Remove useIsomorphicLayoutEffect in Provider

I believe we can actually just remove the effect all together, since it originally in r3f never wanted to be re-run anyway.

bigmistqke committed:
refactor src/index to src/renderer.ts

`core/renderer` was containing code from the deleted `core/reconcile.ts` and a lot of unused code that we deleted in an earlier commit.
The code responsible for rendering the THREE-scene was before this commit located in `solid/index.ts`.

This commit
- moves the types to three-types,
- the rendering-code to `solid/renderer.ts`,
- the empty `core/renderer.ts` is deleted and
- `solid/index.ts` is now only responsible for exporting.

bigmistqke committed:
(fix) import missing DomEvent-type in web/Canvas

bigmistqke committed:
(fix) import `createRoot` from `./renderer` and not solid

bigmistqke committed:
(fix) update import Root in `core/loop

bigmistqke committed:
(fix) rollback updated types Instance and InstanceProps

I had updated `Instance` and `InstanceProps` in commit e0c058d574c7af03a37152bbd342dec14efd0e4a when refactoring with the types of r3f v9 but they substantially changed the typings of Instance.

bigmistqke committed:
(fix) update imports to `../three-types` in `core/utils`

bigmistqke committed:
cleanup exports

bigmistqke committed:
(fix) update package.json

bigmistqke committed:
(cleanup) refactor to explicit type-imports

bigmistqke committed:
(fix) uncomment `useHelper`

bigmistqke committed:
(cleanup) refactor exports

bigmistqke committed:
(cleanup) remove commented out useFrame

bigmistqke committed:
(cleanup) update .prettierrc

.prettierrc before commit was containing the following line
"plugins": ["prettier-plugin-tailwindcss"]
this was causing prettier to not work when the plugin was not present, causing formatting errors throughout the code.

I noticed we were using a different .prettierrc then r3f, this would cause unnecessary differences between the two codebases.
To unify the codestyle of both, I copy+pasted r3f's .prettierrc and ran `prettier --write ./src` from root

bigmistqke committed:
move store.ts to src/solid since it contains solid-code

bigmistqke committed:
replace our custom typing of the three-jsx with r3f's typing

Our branch had custom typing of the jsx-nodes, but since then r3f implemented their own version of it, slightly more streamlined. To unify the codebases I replaced our typing with r3f's.

bigmistqke committed:
remove applyProps from `core/utils` in favor of `solid/renderer`

Before commit there were two applyProps-implementations, one custom made and one from the original r3f-file.

This commit adjusts the custom applyProps function by removing the call to `useThree` and replacing it with `getRootState()`. This way that function can also in the code before the Provider and we can remove r3f's implementation.

bigmistqke committed:
(cleanup) renamed Klass to TSource and defaultProxy to T

and re-written and moved the jsdoc comment for T

bigmistqke committed:
(refactor) remove distinction between `core` and `solid`

bigmistqke committed:
(refactor) add useHelper to utils

bigmistqke committed:
re-write `useHelper` with `Accessor` instead of `createRef`

and remove `createRef`
(I don't really know what useHelper is supposed to be doing tbh)

bigmistqke committed:
(cleanup) remove irrelevant note

bigmistqke committed:
rewrite store-updates as granular updates

bigmistqke committed:
(fix) update import from `./solid/store` to `./core/store` in `src/index.ts

bigmistqke committed:
move context from `./core/context` to `./core/store` like r3f

bigmistqke committed:
(fix/cleanup) imports

bigmistqke committed:
(fix) remove incorrect export

bigmistqke committed:
rename `core/components` to `core/proxy`

a bit of a subjective commit, but imo it reflects better the content and is a better parallel with r3f's `core/reconciler` what this file a replacement is off.

bigmistqke committed:
(cleanup) ignore type-error in `store.set`

bigmistqke committed:
(fix) use argument from produce-callback

bigmistqke committed:
(refactor) to more granular setStore

bigmistqke committed:
(refactor) to more granular setStore

bigmistqke committed:
(fix) update ref-type

bigmistqke committed:
optimization applyProps

Instead of `createRenderEffect(() => applyProps(instanceProps))` we now do `createRenderEffect(mapArray(() => Object.keys(instanceProps), ...)`.
This way we don't have to loop and re-create effects every single time a prop changes.

We also check if a prop has a getter or is a function. If they aren't we only set the property, otherwise we wrap it in an effect.

bigmistqke committed:
(fix) remove `on(() => rootState,` in createEffect

bigmistqke committed:
add TODO

bigmistqke committed:
(cleanup) remove redundant `const state =.rootState

bigmistqke committed:
add flushGlobalEffects from v9

bigmistqke committed:
(cleanup) remove unused `const store` and tidy up

use full names and use similar names as r3f
remove DEBUG-logs

bigmistqke committed:
(cleanup) remove unused code

bigmistqke committed:
implement Portal (still WIP)

bigmistqke committed:
(fix) automatic resizing of canvas

r3f's code had a `store.subscribe(...)` in which they did a diff to check if they should update the camera. With solid we can just place them into a createEffect and let reactivity figure it out. Creates a flickering when resizing, idk if this is a regression or not.

bigmistqke committed:
(refactor) to mimick more closely r3f's structure

bigmistqke committed:
(cleanup) useLoader

Set cache to result instead of callback. In the original code the cache was set to the data and then directly afterwards to the async callback. onProgress is passed through.

bigmistqke committed:
(fix) broken import

and minor cleanup

bigmistqke committed:
(refactor) our code to look more like v9 branch

I copied the contents of r3f's `packages/fiber/src` and used the vscode-extension `Diff Folders` to compare the two directories and made changes so that these two directories look more alike.

Maybe this could be a way forward: have a r3f-directory which contains a clone of `packages/fiber/src` and we manually update them.

bigmistqke committed:
add dts and dts-bundle to dev-dependencies

bigmistqke committed:
add react-three-fiber to git.submodule

bigmistqke committed:
Update Instance-type to r3f's current v9 branch

r3f refactored their internal typings of Instance during https://github.com/pmndrs/react-three-fiber/pull/2465
- before: Instance represented a THREE-object with additional parameters: __r3f: LocalState, children, remove, add and raycast
- now: Instance: LocalState with additional parameter: object, props and children
    - object: is the THREE-object with additional parameter __r3f ( = a circular reference back to this Instance)
    - props contains the props of the component, without children and ref
    - children contain all the children of the component (before children only contained the attached children)

bigmistqke committed:
update example

bigmistqke commented:
I refactored the type of `Instance` as described [in this commit](https://github.com/pmndrs/react-three-fiber/pull/2465)
Now I think our codebase is pretty much aligned with the current v9-branch of r3f.

I also added r3f as a sub-module into the repo, so that it's easy to compare them.

I personally think syncing their repo automatically with ours will be difficult to do, as there are little changes throughout the codebase. Instead I tried to mimmic their codebase as much as possible with ours so we can diff manually (I have been using the vs-extension `Diff Folders` to do that, it's quite handy).

bigmistqke committed:
(refactor) update applyProps to track props[key] only

bigmistqke committed:
(refactor) applyProp(s) to be more in-line with r3f's v9-branch

bigmistqke committed:
export Primitive

bigmistqke committed:
add path-alias to src

bigmistqke committed:
(fix) set .needsUpdate to true if necessary

if key is included in NEEDS_UPDATE
and it changed from falsey to truthy or vice versa.
NEEDS_UPDATE is possibly incomplete.

bigmistqke committed:
(fix) tsconfig playground

bigmistqke committed:
update playground

bigmistqke committed:
(fix) disposal of THREE-object

bigmistqke committed:
(fix) Portal

bigmistqke committed:
add `RenderTarget`-test

bigmistqke committed:
(refactor) update repo to pnpm workspace

repo is now a pnpm workspace
- packages: `@solid-three/fiber` and `@solid-three/drei` (just an empty folder for now)
- apps: `@solid-three/playground`, we could add other apps later one if we want to

the react-three-fiber directory is now moved to `references` and another git-submodule has been added there: `@react-three/drei`

bigmistqke committed:
(fix) remove console.log

bigmistqke committed:
update playground

bigmistqke committed:
(playground) update Tests and added Noop

Noop displays a flaw in our current design:
```
const Noop = () => {
  const box = <T.Mesh>...</T.Mesh>
  return <></>
}
```
will add box to its parent context even if it's not returning the jsx.

bigmistqke committed:
(playground) update examples

bigmistqke committed:
fiber: add `utils/withContext`

a utility function to wrap children in context without having to use JSX.

see https://github.com/solidjs-community/solid-primitives/pull/464

bigmistqke committed:
fiber: create tree with `props.children` instead of `getParent()`-context

Our previous architecture had a design flaw as described 023908c69b9639852e8fbe96f3cede1bd292acc8
```
const Noop = () => {
const box = <T.Mesh>...</T.Mesh>
return <></>
}
```
would add box to its parent, with or without returning the actual jsx.
This was due to the fact that parent-child relationships were established from the child to the parent with context through the `getParent()`-hook.

This commit changes the mechanism from child-to-parent to parent-to-child:
By walking the children from the parent and then attaching/connecting these childs, with a function called `parentChildren`.
This function is used when initializing scene too, the props.children are attached to the THREE.Scene with `parentChildren(() => instance, props)`

Since this would mean that props

bigmistqke committed:
(refactor) rename `makeThreeComponent` to `createThreeComponent`

bigmistqke committed:
fiber: returntype of `<T.Mesh/>` is now `Accessor<THREE.Mesh>`

Before `<T.Mesh/>` return `Accessor<Instance<THREE.Mesh>>` but I think for userland it would be handier if it returned `Accessor<THREE.Mesh>` instead.

That way you could do
```
const mesh = <T.Mesh />
useFrame(() => mesh().rotation.x += 0.1)
return mesh
```
if you don't mind the type-error of `mesh()`

bigmistqke committed:
playground: add tests for Reference

bigmistqke committed:
remove `reference/drei` and add new sub-module in `packages/drei`

this sub-module is a reference to `https://github.com/bigmistqke/drei`. Once there is some progress in drei-port we could move it to `solid-community` too.

bigmistqke committed:
add utility-types

utility-type-imports were causing issues when used in another directory inside the monorepo

bigmistqke committed:
(fix) commit additional files for packages-drei submodule

bigmistqke committed:
Falsey from utils/utility-types`

bigmistqke committed:
fiber: update ref-types

bigmistqke committed:
fiber: update types Primitive

bigmistqke commented:
[for documentation purposes]

The initial way how children were attached to parents, in this branch, was done with the use of `context`, from child to parent: each parent would wrap its children with a `<Context.Provider value={() => Instance}>{props.children}</Context.Provider/>` and each child would check this context with the use of the `useParent`-hook and attach itself to this parent. This however broke the expectation that only returned jsx would attach itself.

```tsx
const Noop = () => {
  const mesh = new <T.Mesh/>
  return </></>
}
```

would actually connect the `mesh` to its parent.

To fix this I adjusted the code (see [b6cbde4](https://github.com/solidjs-community/solid-three/pull/12/commits/b6cbde4ba4609702f24436d429c47dad217e4d61)) to create the hierarchy from parent-to-child instead of from child-to-parent.

Instead of using context we read the `props.children`, from the parent, and attach the children to the parent accordingly. This way only returned jsx becomes a part of the tree.

connorgmeehan commented:
I am looking to use this repo to create a solidjs renderer for pixi js and this maintenance and solid-js refactoring is massively appreciated.   Looking forward to seeing where this PR goes :)

bigmistqke committed:
fiber: add ThreeProps-type

bigmistqke committed:
rename ReactThreeFiber-type to SolidThreeFiber

bigmistqke committed:
fiber: update useLoader-hook to support multiple urls

bigmistqke committed:
fiber: add props-helpers

bigmistqke committed:
fiber: use processProps in Canvas

bigmistqke committed:
fiber: pass <Provider/> straight to children
the memo was causing children to resolve each time it changed.

bigmistqke committed:
fiber: `SolidThree` namespace, remove `Catalogue`

In favor for `SolidThree` namespace.
r3f extends the global namespace of JSX.IntrinsicElements.
In `solid-three` we do not extend the JSX-namespace, but `SolidThree` instead.

bigmistqke committed:
fiber: initial step port to v8 - remove Stages

bigmistqke committed:
fiber: makeshift attachChild-function

bigmistqke committed:
fiber: (fix) remove Stages

bigmistqke committed:
fiber: attach / detach from r3f

bigmistqke committed:
fiber: extend T with SolidThreeElements

bigmistqke committed:
fiber: remove references to `Stages`

bigmistqke committed:
fiber: (fix) .getState
leftover from zustand

bigmistqke committed:
fiber: move `useObject` to inside memo

bigmistqke committed:
fiber: pass `context.Provider` directly into `parentChildren`
instead of using `withContext`

bigmistqke committed:
fiber: delay `useFrame` with 1 frame
this way 99% of all refs will be resolved by then.

bigmistqke committed:
fiber: `T.Suspense` and `createThreeResource`

Inconsistencies between `Suspense` implementation of `react` and `solid` caused `Suspense` to behave differently between `r3f` and `solid-three`:
- in `r3f` a resource throws a promise, this way it is able to defer creation of three-elements until all resources were resolved
- in `solid-three` only `createEffect` and `onMount` are deferred by suspense, and all elements are eagerly created

Besides that `Suspense` was not behaving as expected, where each resource needed to be dealt with on a component-level too.

This commit fixes `Suspense` and creates a custom `T.Suspense` and `createThreeResource`.
The jsx-graph is walked eagerly, and resource-accesses trigger suspenses. This way waterfalls are prevented.
When using `T.Suspense` and `createThreeResource` creation of three-elements are deferred until all resources are resolved.
`createLoader`, and hereby all of `drei`'s loaders too, use `createThreeResource`

bigmistqke committed:
fiber: update package.json

bigmistqke committed:
playground: update playground

bigmistqke committed:
fiber: update `inject` in Portal to be more in line with v8

Before this commit we were only setting a handful of parameters in inject and letting the rest be set by `rootState`.
This was causing bugs, p.ex in `GizmoHelperStories` from `drei`, where the camera of the `Portal` was being set by the `rootState`.
We were missing a couple of loc to filter out the correct keys to inject.

There is for sure a more performant, more solid-like way to merge the `rootState` and the `portalState`, which I will explore in a later commit.

bigmistqke committed:
fiber: `AllConstructorParameters`-typehelper to include overloaded ConstructorParameters

`ConstructorParameters` does not include overloaded constructors
see https://github.com/microsoft/TypeScript/issues/37079 and https://github.com/pmndrs/react-three-fiber/issues/581
`r3f`s solution is to special-case Color in v8 and solve it later for v9, because they faced some problems with the solution/hack described in the typescript github-issue
The hack didn't seem to cause the same issues with us, so I implemented that instead.

It's a manual union of overloaded methods, going up to 7 overloads, which is probably a bit of overkill.

bigmistqke committed:
fiber: seperate `IntrinsicElements` and `IntrinsicComponents`

Before commit we mirrored `r3f` `JSX.IntrinsicElements` with `SolidThree.IntrinsicElements`
`IntrinsicElements` is an interface with the props as values.

An earlier commit added `T.Suspense` and `T.Primitive` to the interface with `Primitive: Parameters<typeof Primitive<T>>[0]` but this didn't work nicely with the generic.
To solve this I added `IntrinsicComponents` to the `SolidThree`-namespace, `Primitive: typeof Primitive` and then in `SolidThreeElements = { ...SolidThree.IntrinsicElements } & { ... SolidThree.IntrinsicComponents

bigmistqke committed:
fiber: use `onMount` instead of `useEffect` in `useFrame`

the subscription-function should only be called once, after suspension is resolved: `onMount` prevents accidental re-subscriptions.

bigmistqke committed:
apps: add `apps/fiber-example`

unported copy+paste from `r3f`

bigmistqke committed:
update package.json and pnpm-lock.yaml

add `build:drei` script

bigmistqke committed:
fiber: rename `constructors` and `components` to `intrinsicElements` and `intrinsicComponents`

bigmistqke commented:
[for documentation purposes]

- we move away from v9 to v8 since it has a more stable api, on advice by @codyjasonbennett.
  - v8: `Stages` are removed
  - v8: loop-functions are scoped inside `createLoop`
  - v9: `Instance` is still typed like v9 [see](https://github.com/pmndrs/react-three-fiber/pull/2465)
  - v9: `T` is still typed dynamically, with custom code to account for overloaded methods [see](https://github.com/solidjs-community/solid-three/pull/12/commits/526497a1e07de2b07473cdb7ce293d71869dc0bb) 
- I have been working on the `drei` port at [solid-drei](https://github.com/bigmistqke/drei) which helped this port too: 
  - `Portal` is fixed
  - `useLoader` has been improved
  - mechanism in place to extend `T` with namespace `SolidThree` and `SolidThree.IntrinsicElements` (for three-elements and `extend()`) and `SolidThree.IntrinsicComponents` (for custom `solid-three` components like `T.Suspense` and `T.Primitive`)
- Improve Suspense:
  - `r3f` has nice loader patterns in which their `useLoader` throws immediately on creation and triggers suspense-boundaries. This way it's always guaranteed that their resources are always loaded. It has the downside that no resources are parallelized (afaik). Solid's way of doing Suspense is fundamentally different, so our previous implementation caused some friction when using `Suspense`.
  - To mimick more closely `r3f` behavior of suspending creation of three-elements until suspended boundary is resolved, while retaining parallel resource loading some work has been done:
    - `<T.Suspense/>`: includes a context with a `{ resolved, addResource }`-value
    - `onSuspense` hook: to read the suspense-context
    - `createThreeResource()`: a wrapper around `createResource` that `context.addResource`)
      - `useLoader` uses `createThreeResource` internally, all of drei's loaders are based on this primitive.
      - `<T.Primitive/>` and `createThreeComponent` both contain code that read `suspenseContext` from `useSuspense` and that can suspend three-element creation in case `!suspenseContext.resolved`.
    - Timeline of creation of a `<T/>`-component or a `<T.Primitive/>`:
      - an initial inspection-run:
        - We do not yet create the three-element 
        - We resolve the children and read the props. 
        - We do not yet connect/attach the children nor apply the props
      - after initial run: 
        - if no resource was read during this initial run:
          - create three-element immediately
          - apply props
        - if a resource was read during this initial run:
          - if component is sibling of a `T.Suspense`-context: 
            - effect/onMount suspended until `Suspense` is resolve 
            - suspend creation until `suspenseContext.resolved`
            - applyProps and manageChildren is suspended until element is created
          - if component is sibling of a regular `Suspense`: 
            - effect/onMount suspended until `Suspense` is resolve
            - create element immediately
            - attach children immediately, but the root is added to the scene only when `Suspense` is resolved
            - apply props with `undefined` for the resource (ignoring the prop), the prop is applied once it is resolved (potentially even before the `Suspense` is resolved) 
          - if component is not a sibling of any type of suspense:
            - create element immediately
            - apply props with `undefined` for the resource (ignoring the prop)
    - When the children are resolved, they go through the same process, causing this implementation to resolve the tree depth-first: we walk to the leafs first and then build from the leafs up to the root, and finally attach the tree to the rest of the scene. Before the implementation the tree was resolved top-down, which was the cause of earlier `Suspense`-inconsistencies where a `<T.Mesh/>` would already be attached to the scene while its `<T.Material/>`-child would have a resource.
  
## Comments: 
  - I am unsure if the whole suspend-element-creation strategy is 100% necessary, with its custom `T.Suspense` and `createThreeElement`, it could very well be an unnecessary complication. The initial inspection-run to collect resources together with regular `Suspense`, resolving the tree from the bottom-up until the Suspense-boundary, could visually be enough. It might even be more performant, since it spreads out the creation of three-elements and the applying of props.
  - We should investigate more what are the actual benefits of suspending three-element creation with all props defined: I think there could be perf benefits if it's for args (I think it could prevent an additional compilation of a shader) but if they are used as methods of an existing three-object I am not sure if there is any perf benefit at all:

```ts
// this could have performance benefits if texture is defined
const material = new THREE.Material({map: texture})

// I am not sure this would, since it would need to compile the shader first anyway
// except if THREE would queueMicrotask the compilation of the shader 🤔
const material = new THREE.Material()
material.map = texture
```
  - This timeline-management comes with the cost of a `createLazyMemo` in `<T.Primitive/>` and all other `<T/>`-components and the double-run means effects need to be run twice (duh). All values are memo'd with `mapArray` so I suppose the extra effects won't be too expensive, but there will be some overhead.

bigmistqke committed:
fiber: update typescript and vite to latest version

bigmistqke committed:
fiber: add `when`-utility

bigmistqke committed:
fiber: initial `inspect`-run

Introduce an initial `inspect`-run in which we resolve children and props, but do not yet create the three-element, attach/connect the children or apply the props.
This way any resource that is present in the new tree gets accessed and their respective `Suspense`s are activated.
The children are resolved depth-first and are then created from the leafs back to the root, until they are finally added to the rest of the scene.

see https://github.com/solidjs-community/solid-three/pull/12#issuecomment-1682258922 for a more detailed description of the timeline of the components.

bigmistqke committed:
fiber: widen PortalProps.container

bigmistqke committed:
fiber: update pnpm-lock.yaml and vite.config.ts

bigmistqke commented:
@connorgmeehan cool! I would be pretty interested if there is a way to generalize this renderer, so you could plug it in other libraries. If you wanna [come and hang out with us at discord](https://discord.com/channels/722131463138705510/1101885290899054662)

bigmistqke committed:
fiber: fix types `<T/>`

A regression happened with extending the SolidThree-namespace where it didn't recognize the newly added T-components.
This was caused bc the type of `ThreeElements` was resolved when the namespace is empty.

bigmistqke committed:
fiber: ts-ignore/fix some type-errors

bigmistqke commented:
I removed all the timeline-management code and the suspension of three-element creation as they didn't compose nicely with ordinary jsx-elements and messed up some expectations, p.ex on when refs are resolved:

```tsx
<T.Mesh ref={mesh}>
  <CustomComponent mesh={mesh}/>
</T.Mesh>
``` 

In the above example `CustomComponent` expects to receive `mesh` immediately, but with the initial inspection-run this resulted into undefined.

bigmistqke committed:
fiber: let `useThree` take a selector

bigmistqke committed:
fiber: remove scheduling code from `proxy`

bigmistqke committed:
candycrush: initial commit

bigmistqke committed:
candycrush: update candycrush

rotate matrix
add envMap to gems

bigmistqke committed:
fiber: `getObject()?.dispose?.()`

was creating issues with drei's `<OrbitControls/>`

bigmistqke committed:
tetris: initial commit

bigmistqke committed:
tetris: add dropped tetromino-preview

bigmistqke committed:
fiber: port tests to solid and copy shared utils

bigmistqke committed:
fiber: add Portal to T

bigmistqke committed:
fiber: add `get` back to RootState

a bit of a silly property, since with solid-store we can just access the store directly, but handy when porting.

bigmistqke committed:
fiber: preserve reactivity in props passed to configure

Use solid's utility functions `mergeProps` and `splitProps` to compose arguments, instead of de-structuring and spreading.
This way reactivity is preserved.

bigmistqke committed:
fiber: pass `Accessor<JSX.Element>` to root.render instead of `{ children: JSX.Element }`

bigmistqke committed:
fiber: eagerly create Portal instead of waiting for mount.

still remnant of experiment with suspending creation in case of Suspense

bigmistqke committed:
workspace: update order workspace-packages

bigmistqke committed:
workspace: add postinstall to build `packages/*`

bigmistqke committed:
workspace: preinstall `git submodule update --remote`

bigmistqke committed:
workspace: add `pnpm:devPreinstall`
bc npm lifecycles are the worst.

bigmistqke committed:
workspace: add `build:packages` and `install:packages`
and add complex git-command copy+pasted from
https://stackoverflow.com/a/62192503/4366929
otherwise submodule's head would be detached
after `git submodule update --remote --rebase`

bigmistqke committed:
workspace: import packages with `workspace:*`

bigmistqke committed:
workspace: add concurrently to devdependencies

bigmistqke committed:
chore: update pnpm-lock.yaml

bigmistqke committed:
workspace: rename `pnpm:devPreinstall` to `prepare`

devPreinstall still created a bug where `apps/playground` would error
bc it wasn't able to read the package.json of drei (although
the submodule was already cloned)

bigmistqke committed:
workspace: add solid-xr to submodules

bigmistqke committed:
workspace: update pnpm-lock.yaml

bigmistqke committed:
workspace: add `@types/jest` to devDeps

bigmistqke committed:
workspace: update pnpm-lock.yaml

bigmistqke committed:
playground: update example

bigmistqke committed:
workspace: add jest to tsconfig

bigmistqke committed:
race: add race-example

bigmistqke committed:
race: (fix) set path drei to `workspace:*`

connorgmeehan commented:
> @connorgmeehan cool! I would be pretty interested if there is a way to generalize this renderer, so you could plug it in other libraries. If you wanna [come and hang out with us at discord](https://discord.com/channels/722131463138705510/1101885290899054662)

Sorry for the late response.  It's funny you say that.  That's one of the first things that I did when adapting this code.  

Here's an early version of the generalised proxy renderer. I haven't really bothered open sourcing it the up to date code because I'm working in a private monorepo with my app.
https://github.com/sanspointes/constructables/
Basically it just lets you create a new root with a context and lets you wrap any classes into a component with automatic prop typing (which is honestly pretty buggy and slow for the typescript compiler).  It also prefers defining implied props explicitly (such as `position-x: number`) rather than at runtime so you get proper typing.
Also here's the [pixi renderer implementation](https://github.com/sanspointes/solixi) that uses the generalised renderer.

I am running into performance issues, especially surrounding  the way props are tracked when components are created.  I would love to combine efforts. :)

Also can you re-share the discord link? I can't open it.

bigmistqke commented:
continue development in `next`-branch
